components/tidb_query_datatype/src/codec/data_type/chunked_vec_enum.rs

```diff
@@ -0,0 +1,142 @@
+use super::bit_vec::BitVec;
+use super::{ChunkRef, ChunkedVec, UnsafeRefInto};
+use super::{Enum, EnumRef};
+use crate::impl_chunked_vec_common;
+use std::sync::Arc;
+use tikv_util::buffer_vec::BufferVec;
+
+#[derive(Debug, Clone)]
+pub struct ChunkedVecEnum {
+    data: Arc<BufferVec>,
+    bitmap: BitVec,
+    // MySQL Enum is 1-based index, value == 0 means this enum is ''
+    value: Vec<usize>,
+}
+
+impl ChunkedVecEnum {
+    impl_chunked_vec_common! { Enum }
+
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            data: Arc::new(BufferVec::new()),
+            bitmap: BitVec::with_capacity(capacity),
+            value: Vec::with_capacity(capacity),
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.value.len()
+    }
+
+    #[inline]
+    pub fn push_data(&mut self, value: Enum) {
+        self.bitmap.push(true);
+        self.value.push(value.value());
+    }
+
+    #[inline]
+    pub fn push_null(&mut self) {
+        self.bitmap.push(false);
+        self.value.push(0);
+    }
+
+    pub fn truncate(&mut self, len: usize) {
+        if len < self.len() {
+            self.bitmap.truncate(len);
+            self.value.truncate(len);
+        }
+    }
+
+    pub fn capacity(&self) -> usize {
+        self.bitmap.capacity().max(self.value.capacity())
+    }
+
+    pub fn append(&mut self, other: &mut Self) {
+        self.value.append(&mut other.value);
+        self.bitmap.append(&mut other.bitmap);
+    }
+
+    #[inline]
+    pub fn get(&self, idx: usize) -> Option<EnumRef> {
+        assert!(idx < self.len());
+        if self.bitmap.get(idx) {
+            Some(EnumRef::new(&self.data, *self.value.get(idx).unwrap()))
+        } else {
+            None
+        }
+    }
+
+    pub fn to_vec(&self) -> Vec<Option<Enum>> {
+        let mut x = Vec::with_capacity(self.len());
+        for i in 0..self.len() {
+            x.push(if self.bitmap.get(i) {
+                Some(Enum::new(self.data.clone(), *self.value.get(i).unwrap()))
+            } else {
+                None
+            });
+        }
+        x
+    }
+}
+
+impl PartialEq for ChunkedVecEnum {
+    fn eq(&self, other: &Self) -> bool {
+        if self.data.len() != other.data.len() {
+            return false;
+        }
+        for idx in 0..self.data.len() {
+            if self.data[idx] != other.data[idx] {
+                return false;
+            }
+        }
+
+        if !self.bitmap.eq(&other.bitmap) {
+            return false;
+        }
+
+        if !self.value.eq(&other.value) {
+            return false;
```

Comment:
Is it possible that one position is `None`, but their stored value is different? I suggest compare them one element by one element using `EnumRef`. At the same time, I guess other `ChunkedVec` may also suffer from this issue.

Reply:
I think this is a bug. If you have time, you may also help implement correct `PartialEq` on other chunked vector types.

Reply:
I didn't get it, is `but their stored value is different` possible?

Reply:
Let take a look at this example. For `[Some(1), Some(2), None]`, we could have different stored representation.

`Bitmap: 110, Data: 1, 2, 0`
`Bitmap: 110, Data: 1, 2, 1`
`Bitmap: 110, Data: 1, 2, 3`

Are the same. As bitmap for the third position is `0`, its corresponding value is `None`, no matter what is stored in value. But current `PartialEq` will think they are different.

Reply:
I thought the compare only happend while `data` is the same?

Reply:
I've made a typo. `Data` should be `Value`.

Reply:
We will operate `bitmap` and `value`  together, there should not be a situation where `bitmap` is 0 and `value` is some other value. Am I missing something?

Reply:
You're right. If value is always 0 when bitmap is 0, then there's no such issue. Maybe we should document this behavior into docs of `ChunkedVec`, to warn future developers ensuring this property?

Reply:
OK, I will document this.

Reply:
Add comments to `ChunkedVec` need a seperate PR?

Reply:
Yes, you may just add it for `Enum` for now

Reply:
OK